crates/router/src/services/refresh_token.rs

```diff
@@ -0,0 +1,151 @@
+        address: router_data.address,
+        amount_captured: router_data.amount_captured,
+        auth_type: router_data.auth_type,
+        connector: router_data.connector,
+        connector_auth_type: router_data.connector_auth_type,
+        connector_meta_data: router_data.connector_meta_data,
+        description: router_data.description,
+        router_return_url: router_data.router_return_url,
+        payment_id: router_data.payment_id,
+        payment_method: router_data.payment_method,
+        payment_method_id: router_data.payment_method_id,
+        return_url: router_data.return_url,
+        status: router_data.status,
+        attempt_id: router_data.attempt_id,
+        access_token: router_data.access_token,
+    }
+}
+
+pub async fn add_access_token<
+    F: Clone + 'static,
+    Req: Debug + Clone + 'static,
+    Res: Debug + Clone + 'static,
+>(
+    state: &AppState,
+    connector: &api_types::ConnectorData,
+    merchant_account: &storage::MerchantAccount,
+    router_data: &types::RouterData<F, Req, Res>,
+) -> RouterResult<types::AddAccessTokenResult> {
+    if connector.connector_name.supports_access_token() {
+        let merchant_id = &merchant_account.merchant_id;
+        let db = &*state.store;
+        let old_access_token = db
+            .get_access_token(merchant_id, connector.connector.id())
+            .await
+            .change_context(errors::ApiErrorResponse::InternalServerError)
+            .attach_printable("DB error when accessing the access token")?;
+
+        let res = match old_access_token {
+            Some(access_token) => Ok(Some(access_token)),
+            None => {
+                let cloned_router_data = router_data.clone();
+                let refresh_token_request_data =
+                    types::AccessTokenRequestData::from(router_data.connector_auth_type.clone());
+                let refresh_token_response_data: Result<types::AccessToken, types::ErrorResponse> =
+                    Err(types::ErrorResponse::default());
+                let refresh_token_router_data =
+                    router_data_type_conversion::<_, api_types::AccessTokenAuth, _, _, _, _>(
+                        cloned_router_data,
+                        refresh_token_request_data,
+                        refresh_token_response_data,
+                    );
+                refresh_connector_auth(
+                    state,
+                    connector,
+                    merchant_account,
+                    &refresh_token_router_data,
+                )
+                .await?
+                .async_map(|access_token| async {
+                    //Store the access token in db
+                    let db = &*state.store;
+                    // This error should not be propagated, we don't want payments to fail once we have
+                    // the access token, the next request will create new access token
+                    let _ = db
```

Comment:
let _ is not required

Reply:
let _ is required or else cargo will complain about result not being used